### PR TITLE
Fix hqq parser (ported from LordVenom/venom-xbmc-addons)

### PR DESCRIPTION
**Pull request by jose1711**
had to go with my own get methods as replies are coming as http/404

---

This change is [Reviewable](https://reviewable.io/reviews/kodi-czsk/script.module.stream.resolver/33)

### CODE REVIEW

File: lib/server/hqqresolver.py [hunk 1]
```diff
@@ -8,18 +8,45 @@
...
+    req.add_header('User-Agent', UA)
+    try:
+        response = urllib2.urlopen(req)
+        data = response.read()
+        response.close()
+    except urllib2.HTTPError, error:
+        data=error.read()
+    print('len(data) %s' % len(data))
+    return data
+
```
Comment (lubo) [inline lib/server/hqqresolver.py]: Why have you copied the code from `util` ? I'm not OK with reimplementing these functions.

File: lib/server/hqqresolver.py [hunk 2]
```diff
@@ -135,22 +162,74 @@ def resolve(url):
...
+                                'b' : '1',
+                                'vid' : vid
+                               }
+                    data = request("http://hqq.tv/player/get_md5.php?" + \
+                           urllib.urlencode(get_data), headers)
+                    jsondata = json.loads(data)
+                    encodedm3u = jsondata['file']
+                    decodedm3u=_decode2(encodedm3u.replace('\\', ''))
+
+                    return [{'url': decodedm3u+'|User-Agent=Mozilla/5.0 (iPhone; CPU iPhone OS 5_0_1 like Mac OS X)', 'quality': '360p'}]
```
Comment (lubo) [inline lib/server/hqqresolver.py]: This line is too long, trim it to maximum of 100 characters.

File: lib/server/hqqresolver.py (hunk 1 above)
Comment (jose1711) [inline lib/server/hqqresolver.py]: this is by no means a 1:1 copy. please checkout the following example:

```
>>> url='http://hqq.tv/sec/player/embed_player.php...'
>>> req = urllib2.Request(url)
>>> req.add_header('User-Agent', 'Mozilla/6.0 (Windows; U; Windows NT 5.1; en-GB; rv:1.9.0.5) Gecko/2008092417 Firefox/3.0.3')
>>> try:
...  urllib2.urlopen(req)
... except urllib2.HTTPError, error:
...  len(error.read())
15588
```

i don't see any way around this as the parsed data must come from the error response.

File: lib/server/hqqresolver.py (hunk 2 above)
Comment (jose1711) [inline lib/server/hqqresolver.py]: done

File: lib/server/hqqresolver.py (hunk 1 above)
Comment (lubo) [inline lib/server/hqqresolver.py]: What about implementing the error handling in `util` ?

File: lib/server/hqqresolver.py [hunk 3]
```diff
@@ -135,22 +162,75 @@ def resolve(url):
...
+            _local7 = 0
+            while _local7 < len(_local4):
+                if _local3[_local7 + 1] == 64:
+                    break
+                _local2 += chr(_local4[_local7])
+                _local7 += 1
+            _local5 += 4
+        return _local2
+
+    return _xc13(K12K(file_url, 'e'))
```
Comment (lubo) [inline lib/server/hqqresolver.py]: Why did you duplicate this function ?

Comment (jose1711) [inline lib/server/hqqresolver.py]: yeah, my mistake. fixed

File: lib/server/hqqresolver.py [hunk 4]
```diff
@@ -8,17 +8,19 @@
...
-import re
 import util
+import re
 import base64
 import urllib
+import urllib2
 
 __name__ = 'hqq'
-
+UA = 'Mozilla/6.0 (Windows; U; Windows NT 5.1; en-GB; rv:1.9.0.5) Gecko/2008092417 Firefox/3.0.3'
```
Comment (lubo) [inline lib/server/hqqresolver.py]: This line is no longer necessary.

File: lib/server/hqqresolver.py [hunk 5]
```diff
@@ -161,7 +172,7 @@ def _regex(url):
         return match
     match = re.search(r'(hqq|netu)\.tv/player/hash\.php\?hash=\d+', url)
     if match:
-        match = re.search(r'var\s+vid\s*=\s*\'(?P<vid>[^\']+)\'', urllib.unquote(util.request(url)))
+        match = re.search(r'var\s+vid\s*=\s*\'(?P<vid>[^\']+)\'', urllib.unquote(request(url)))
```
Comment (lubo) [inline lib/server/hqqresolver.py]: `request()` not longer exists.

File: lib/util.py [hunk 6]
```diff
@@ -77,24 +77,30 @@ def cache_cookies(cache):
         cache.set(CACHE_COOKIES, _cookie_jar.dump())
 
 
-def request(url, headers={}):
+def request(url, headers={}, errors_are_useful=False):
```
Comment (lubo) [inline lib/util.py]: Do we really need this extra argument ? I'd read the error every time it occurs, would solve some crashes in other resolvers and plugins.

File: lib/util.py [hunk 7]
```diff
@@ -77,24 +77,30 @@ def cache_cookies(cache):
...
     debug('request: %s' % url)
     req = urllib2.Request(url, headers=headers)
     req.add_header('User-Agent', UA)
-    response = urllib2.urlopen(req)
-    data = response.read()
-    response.close()
+    try:
+        response = urllib2.urlopen(req)
+        data = response.read()
+        response.close()
```
Comment (lubo) [inline lib/util.py:87]: Put this line under `finally` clause, so the response get closed every time.

File: lib/util.py [hunk 8]
```diff
@@ -77,24 +77,30 @@ def cache_cookies(cache):
...
+    except urllib2.HTTPError, error:
+        if errors_are_useful: data=error.read()
```
Comment (Kuci) [inline lib/util.py]: Everything mentioned above applies to this function as well.

File: lib/server/hqqresolver.py [hunk 9]
```diff
@@ -136,22 +138,31 @@ def resolve(url):
                 for idx in range(len(data)):
                     post_data[data[idx][0]] = data[idx][1]
                 data = urllib.unquote(util.request("http://hqq.tv/sec/player/embed_player.php?" +
-                                                   urllib.urlencode(post_data), headers))
-                vid_server = re.search(r'var\s*vid_server\s*=\s*"([^"]*?)"', data)
-                vid_link = re.search(r'var\s*vid_link\s*=\s*"([^"]*?)"', data)
+                                                   urllib.urlencode(post_data), headers, errors_are_useful=True))
+                servervarname=re.search('server_1: ([^,]+)',re.findall('md5.*',data)[0]).group(1)
+                linkvarname=re.search('link_1: ([^,]+)',re.findall('md5.*',data)[0]).group(1)
```
Comment (lubo) [inline lib/server/hqqresolver.py]: You should use raw strings here.

File: lib/util.py (hunk 8 above)
Comment (Kuci) [inline lib/util.py]: Everything mentioned above applies to this function as well.

File: lib/util.py (hunk 7 above)
Comment (jose1711) [inline lib/util.py:87]: if i do so then response is not populated and i get:
UnboundLocalError: local variable 'response' referenced before assignment

File: lib/server/hqqresolver.py (hunk 4 above)
Comment (jose1711) [inline lib/server/hqqresolver.py]: thank you for your relentless review - hopefully all suggestions are implemented

File: lib/util.py (hunk 7 above)
Comment (lubo) [inline lib/util.py:87]: I see, my bad. Maybe you should put `error.close()` into except clause.

File: lib/server/hqqresolver.py [hunk 10]
```diff
@@ -8,18 +8,19 @@
...
+# *  update based on https://github.com/LordVenom/
 # */
 from StringIO import StringIO
 import json
-import re
 import util
+import re
 import base64
 import urllib
+import urllib2
```
Comment (lubo) [inline lib/server/hqqresolver.py]: urllib2 is not longer used.

File: lib/server/hqqresolver.py [hunk 11]
```diff
@@ -137,21 +138,30 @@ def resolve(url):
                     post_data[data[idx][0]] = data[idx][1]
                 data = urllib.unquote(util.request("http://hqq.tv/sec/player/embed_player.php?" +
                                                    urllib.urlencode(post_data), headers))
-                vid_server = re.search(r'var\s*vid_server\s*=\s*"([^"]*?)"', data)
-                vid_link = re.search(r'var\s*vid_link\s*=\s*"([^"]*?)"', data)
+                servervarname=re.search('server_1: ([^,]+)',re.findall(r'md5.*',data)[0]).group(1)
+                linkvarname=re.search('link_1: ([^,]+)',re.findall(r'md5.*',data)[0]).group(1)
```
Comment (lubo) [inline lib/server/hqqresolver.py]: You should put raw strings to `re.search()` too.

**Pull request merged**